sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala

```diff
@@ -40,10 +40,17 @@ case class Average(child: Expression) extends DeclarativeAggregate with Implicit
 
   override def children: Seq[Expression] = child :: Nil
 
-  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType)
+  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, DateType)
```

Comment:
@Fokko, before you go further, can we check other DBMSes as references? I would like to avoid having a variant behaviour in Spark alone compared to other DBMSes ...

Reply:
Sure, that makes sense. See the details below, let me know if I'm missing something, but I don't think there is a real consensus on the subject.

# Postgres

For postgres, it is just unsupported

```
postgres@366ecc8a0fb9:/$ psql
psql (12.3 (Debian 12.3-1.pgdg100+1))
Type "help" for help.

postgres=# SELECT CAST(CAST('2020-01-01' AS DATE) AS decimal);
ERROR:  cannot cast type date to numeric
LINE 1: SELECT CAST(CAST('2020-01-01' AS DATE) AS decimal);
               ^

postgres=# SELECT CAST(CAST('2020-01-01' AS DATE) AS integer);
ERROR:  cannot cast type date to integer
LINE 1: SELECT CAST(CAST('2020-01-01' AS DATE) AS integer);
               ^

The way to get the epoch in days is:

postgres=# SELECT EXTRACT(DAYS FROM (now() - '1970-01-01'));
date_part 
-----------
    18422
(1 row)
```

# MySQL

For MySQL it will convert it automatically to a YYYYMMDD format:

```
mysql> SELECT CAST(CAST('2020-01-01' AS DATE) AS decimal);
+---------------------------------------------+
| CAST(CAST('2020-01-01' AS DATE) AS decimal) |
+---------------------------------------------+
|                                    20200101 |
+---------------------------------------------+
1 row in set (0.00 sec)
```

Converting to an int is not allowed:

```
mysql> SELECT CAST(CAST('2020-01-01' AS DATE) AS int);
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'int)' at line 1

mysql> SELECT CAST(CAST('2020-01-01' AS DATE) AS bigint);
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'bigint)' at line 1
```

# BigQuery

Unsupported

![image](https://user-images.githubusercontent.com/1134248/84114035-ede8e400-aa2b-11ea-9c0c-d0764164d549.png)

https://cloud.google.com/bigquery/docs/reference/standard-sql/conversion_rules

# Excel

The greatest DBMS of them all:
![image](https://user-images.githubusercontent.com/1134248/84115139-d4489c00-aa2d-11ea-98b3-375f1d0c6098.png)

Which is the epoch since 01-01-1900 :)

Reply:
For Avro it is milliseconds since epoch:
https://github.com/apache/avro/blob/master/lang/java/avro/src/main/java/org/apache/avro/reflect/DateAsLongEncoding.java

For Parquet it is days since epoch:
https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#date

Also ORC is based around days since Epoch:
https://github.com/apache/orc/blob/master/java/core/src/java/org/threeten/extra/chrono/HybridDate.java

Also with this, we keep parity with the Catalyst type :)

Reply:
@HyukjinKwon what are your thoughts on this? Can we move this forward?